### PR TITLE
[SPORK-70] Incremental Zobrist hash and use transposition table in Quiescence search

### PR DESCRIPTION
**Pull request by KYLChiu**
* Implements incremental Zobrist hash as opposed to computing the hash via full information.
* Added transposition table usage in Quiescence search.

This leads the performance of transposition table to be fairly equal/slightly better to base case:

FEN: 8/8/8/8/5R2/2pk4/5K2/8 b - - 0 1 (depth 6)
Base, 20302099 function calls (20229039 primitive calls) in 6.205 seconds
Transposition table, 11647835 function calls (11604146 primitive calls) in 4.043 seconds

FEN: r1b2rk1/ppqn1pbp/6p1/3pp3/7B/2PB1N2/PP3PPP/R2QR1K1 w - - 0 16 (depth 6)
Base, 48916378 function calls (48808919 primitive calls) in 17.582 seconds
TT, 44415536 function calls (44317989 primitive calls) in 17.193 seconds 

However at higher depths:
FEN: r1b2rk1/ppq2ppp/3bpn2/3pP3/8/2PB2B1/PP1N1PPP/R2QK2R b KQ - 0 11  (depth 7):
Base, 361472569 function calls (360480492 primitive calls) in 130.606 seconds
TT, 372255883 function calls (371297000 primitive calls) in 141.766 seconds

 I am not quite happy with it yet, but would like this to go in as this is a vast improvement over full hash computation. Will explore avenues of improvement of TT and optimizing the hash calculation in next PR.

### CODE REVIEW

File: tests/test_searcher.py [hunk 1]
```diff
@@ -149,63 +149,63 @@ def test_perf_transposition_table(
...
-        """Performance test with combined general performance config on"""
-        self._run_perf_analytics(
-            request_fixture.node.name,
-            fen=fen_string,
-            max_depth=max_depth,
-            enable_null_move_pruning=True,
-            enable_delta_pruning=True,
-            enable_aspiration_windows=True,
-        )
+    # @pytest.mark.slow
```
Comment (KYLChiu) [inline tests/test_searcher.py:152]: Oops, I'll fix these

**Comment by ccjeremylo** [comment]
> * Implements incremental Zobrist hash as opposed to computing the hash via full information.
> * Added transposition table usage in Quiescence search.
> * Cleanup runner
> * Fix docstrings of runner and transposition table.
> 
> This leads the performance of transposition table to be fairly equal/slightly better to base case:
> 
> FEN: 8/8/8/8/5R2/2pk4/5K2/8 b - - 0 1 (depth 6) Base, 20302099 function calls (20229039 primitive calls) in 6.205 seconds Transposition table, 11647835 function calls (11604146 primitive calls) in 4.043 seconds
> 
> FEN: r1b2rk1/ppqn1pbp/6p1/3pp3/7B/2PB1N2/PP3PPP/R2QR1K1 w - - 0 16 (depth 6) Base, 48916378 function calls (48808919 primitive calls) in 17.582 seconds TT, 44415536 function calls (44317989 primitive calls) in 17.193 seconds
> 
> However at higher depths: FEN: r1b2rk1/ppq2ppp/3bpn2/3pP3/8/2PB2B1/PP1N1PPP/R2QK2R b KQ - 0 11 (depth 7): Base, 361472569 function calls (360480492 primitive calls) in 130.606 seconds TT, 372255883 function calls (371297000 primitive calls) in 141.766 seconds
> 
> I am not quite happy with it yet, but would like this to go in as this is a vast improvement over full hash computation. Will explore avenues of improvement of TT and optimizing the hash calculation in next PR.

I believe TT is switched on by default right?

**Comment by KYLChiu** [comment]
> > * Implements incremental Zobrist hash as opposed to computing the hash via full information.
> 
> > * Added transposition table usage in Quiescence search.
> 
> > * Cleanup runner
> 
> > * Fix docstrings of runner and transposition table.
> 
> > 
> 
> > This leads the performance of transposition table to be fairly equal/slightly better to base case:
> 
> > 
> 
> > FEN: 8/8/8/8/5R2/2pk4/5K2/8 b - - 0 1 (depth 6) Base, 20302099 function calls (20229039 primitive calls) in 6.205 seconds Transposition table, 11647835 function calls (11604146 primitive calls) in 4.043 seconds
> 
> > 
> 
> > FEN: r1b2rk1/ppqn1pbp/6p1/3pp3/7B/2PB1N2/PP3PPP/R2QR1K1 w - - 0 16 (depth 6) Base, 48916378 function calls (48808919 primitive calls) in 17.582 seconds TT, 44415536 function calls (44317989 primitive calls) in 17.193 seconds
> 
> > 
> 
> > However at higher depths: FEN: r1b2rk1/ppq2ppp/3bpn2/3pP3/8/2PB2B1/PP1N1PPP/R2QK2R b KQ - 0 11 (depth 7): Base, 361472569 function calls (360480492 primitive calls) in 130.606 seconds TT, 372255883 function calls (371297000 primitive calls) in 141.766 seconds
> 
> > 
> 
> > I am not quite happy with it yet, but would like this to go in as this is a vast improvement over full hash computation. Will explore avenues of improvement of TT and optimizing the hash calculation in next PR.
> 
> 
> 
> I believe TT is switched on by default right?
> 
Off by default, the performance stats don't justify it yet

File: tests/test_searcher.py (hunk 1 above)
Comment (KYLChiu) [reply tests/test_searcher.py]: Done.

File: scripts/container-postCreate.sh [hunk 2]
```diff
@@ -9,5 +9,6 @@ pip install -r requirements.txt
 git config --global --add safe.directory $(pwd)
 pre-commit install --hook-type pre-commit
 pre-commit install --hook-type pre-push
+mypy --install-types --non-interactive
```
Comment (ccjeremylo) [inline scripts/container-postCreate.sh:12]: Sorry super basic question - can you remind me how do we use this bash script?

File: sporkfish/searcher/minimax.py [hunk 3]
```diff
@@ -129,7 +129,14 @@ def _aspiration_windows_search(
             )
         return score, move
 
-    def _quiescence(self, board: Board, depth: int, alpha: float, beta: float) -> float:
+    def _quiescence_search(
```
Comment (ccjeremylo) [inline sporkfish/searcher/minimax.py:132]: Nice rename

[nitpick/oppinion] for me a `search` method would return a `chess.Move`, perhaps `_quiescence_score` would be a better name?

File: sporkfish/searcher/negamax.py [hunk 4]
```diff
@@ -39,49 +41,75 @@ def _negamax(
         """
         value = -float("inf")
```
Comment (ccjeremylo) [inline sporkfish/searcher/negamax.py:43]: It is my fault tbh, but this function is way to big (>100 lines long), very hard to read...

Happy to refactor it later myself...(refactoring [sparks joy](https://konmari.com/marie-kondo-rules-of-tidying-sparks-joy/) for me tbf)

According to [Uncle Bob](https://softwareengineering.stackexchange.com/questions/133404/what-is-the-ideal-length-of-a-method-for-you#:~:text=In%20Clean%20Code%3A%20A%20Handbook,not%20be%20100%20lines%20long.) - functions should rarely by more than 20 lines (esp when we are already writing in python lol)

**Review by ccjeremylo — COMMENTED** [review]
I got to go sorry - publishing some comments here first, I'd say I have reviewed around 20% of the change only...probably won't be able to continue reviewing until tomorrow afternoon

File: README.md [hunk 5]
```diff
@@ -68,7 +70,7 @@ This generates an interactive bash shell for you to run the program in.
 Check out the bot on lichess [here](https://lichess.org/@/Sporkfish)! To run the bot, create a file in the root directory named `api_token.txt`. Add your Lichess bot API token. Then run:
 
 ```
-python3 main.py
+python3 -O main.py
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply README.md:73]: This removes asserts. For debugging we can use the name python3 main.py.

File: scripts/container-postCreate.sh (hunk 2 above)
Comment (KYLChiu) [reply scripts/container-postCreate.sh:12]: devcontainer runs this after it builds the container.

File: sporkfish/searcher/minimax.py (hunk 3 above)
Comment (KYLChiu) [reply sporkfish/searcher/minimax.py:132]: Ok, I'll just revert back to the old name later - negamax_score doesn't sound great imo.

File: sporkfish/searcher/negamax.py (hunk 4 above)
Comment (KYLChiu) [reply sporkfish/searcher/negamax.py:43]: I suspect it wont be as easy, but please do if it sparks joy :)

File: sporkfish/searcher/minimax.py [hunk 6]
```diff
@@ -195,6 +202,13 @@ def _quiescence(self, board: Board, depth: int, alpha: float, beta: float) -> fl
         :rtype: float
         """
 
+        # Probe the transposition table for an existing entry
+        # We treat all cases as depth 0, so essentially as an static evaluation
```
Comment (ccjeremylo) [inline sporkfish/searcher/minimax.py:206]: Sorry can you elaborate this inline comment?

File: sporkfish/searcher/negamax.py [hunk 7]
```diff
@@ -38,49 +40,76 @@ def _negamax(
...
+                    zobrist_state,
+                    previous_piece_from_square,  # type: ignore
+                    captured_piece,
+                )
+                if zobrist_state
+                else None
+            )
+
+            child_value = -self._negamax(
+                board, depth - 1, -beta, -alpha, child_zobrist_state
```
Comment (ccjeremylo) [inline sporkfish/searcher/negamax.py:110]: 👍

File: sporkfish/transposition_table.py [hunk 8]
```diff
@@ -1,40 +1,49 @@
...
         score: float,
     ) -> None:
         """
         Store an entry in the transposition table.
-
-        Parameters:
-        - zobrist_hash (int): The Zobrist hash value for the board position.
-        - depth (int): The depth at which the score was calculated.
-        - score (float): The score associated with the board position.
+        Only stores if the existing entry depth is lower than the input one.
```
Comment (ccjeremylo) [inline sporkfish/transposition_table.py:24]: Nice change - ty for cleaning up inconsistent docstring styles

File: sporkfish/transposition_table.py [hunk 9]
```diff
@@ -1,40 +1,49 @@
...
+
 
 class TranspositionTable:
-    def __init__(self, dct: Dict[int, Dict[str, float]]) -> None:
+    def __init__(self, dct: Dict[np.int64, Dict[str, float]]) -> None:
         """
-        Initialize the TranspositionTable object, wrapping a shared_dict.
+        Initialize the TranspositionTable object.
 
+        :param dct: A dictionary containing Zobrist hash keys and associated entries.
```
Comment (ccjeremylo) [inline sporkfish/transposition_table.py:11]: Very helpful description

File: sporkfish/transposition_table.py (hunk 8 above)
Comment (ccjeremylo) [reply sporkfish/transposition_table.py:24]: This comment seems to contradict with the code subtlety?

The code:
```
if not existing_entry or depth > existing_entry["depth"]:
```
My interpretation of the comment `Only stores if the existing entry depth is lower than the input one`:
```
if not existing_entry and depth > existing_entry["depth"]:
```

I suspect and `and` clause is what you wanted? As you only want to overwrite existing entries when we are visiting higher level (nearer to the top) node? 

I might be understanding this wrongly - lmk

**Review by ccjeremylo — COMMENTED** [review]
Seems very well designed - LGTM - left some comments.

Did you (by any chance) did any perf analysis? I noticed you decided to switch this feature on, keen to know the rationale

File: config.yml [hunk 10]
```diff
@@ -10,7 +10,7 @@ SearcherConfig:
   enable_null_move_pruning: True # flag to enable null move pruning, i.e. checking for a beta-cutoff by passing our turn
   enable_futility_pruning: False # flag to enable futility_pruning, i.e. prune quiet moves that do not raise alpha in negamax search.
   enable_delta_pruning: True # flag to enable delta pruning, i.e. prune capturing moves that do not raise alpha in quiescence search.
-  enable_transposition_table: False # flag to enable position cache, so positions are not evaluated twice
+  enable_transposition_table: True # flag to enable position cache, so positions are not evaluated twice
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply config.yml:13]: I felt like it wasn't much of a loss to do so - we don't expect to hit much higher than depth 7 atm and at depth 6 the perf is similar.

File: sporkfish/zobrist_hasher.py [hunk 11]
```diff
@@ -1,36 +1,182 @@
+from dataclasses import dataclass
+from typing import Optional
+
 import chess
 import numpy as np
 from numba import njit
 
 from sporkfish.board.board import Board
 
+_int64_min_val = np.iinfo(np.int64).min
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:10]: I forgot - oops - good catch thanks!

File: sporkfish/zobrist_hasher.py [hunk 12]
```diff
@@ -1,36 +1,182 @@
...
+class ZobristHasher:
+    """
+    This allows caching via the transposition table, so we don't have to evaluate positions twice.
+    Two methods are provided:
+    - One hashes the board statically, doing this by retrieving the full board and hence is slow.
+    - The other hashes the board after a move is made (i.e. incrementally), and is designed to be faster.
+    """
+
+    @staticmethod
+    def _parse_ep_file(board: Board) -> np.int8:
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:152]: No particular reason, it doesn't need self plus doesn't need to be called outside

File: sporkfish/zobrist_hasher.py [hunk 13]
```diff
@@ -1,36 +1,182 @@
...
+    board_hash = _en_passant_hash(board_hash, curr_en_passant_file)
 
-class BoardInfo:
+    board_hash = _castling_hash(board_hash, prev_castling_rights)
+    board_hash = _castling_hash(board_hash, curr_castling_rights)
 
+    return board_hash  # type: ignore
+
+
+@dataclass
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:125]: Yes

File: sporkfish/zobrist_hasher.py [hunk 14]
```diff
@@ -40,82 +186,93 @@ def __init__(self, board: Board) -> None:
...
-        Parameters:
-        - board (Board): The chess board.
+        squares = np.array(squares_list, dtype=np.int8)
+        colored_piece_types = np.array(colored_piece_types_list, dtype=np.int8)
 
-        Returns:
-        - int: The computed Zobrist hash value for the board.
-        """
+        ep_file = ZobristHasher._parse_ep_file(board)
+        castling_rights = ZobristHasher._parse_castling_rights(board)
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:267]: Yes

File: config.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (ccjeremylo) [reply config.yml:13]: Have we figured out why perf is similar? In theory (naively thinking), TT/incremental ZH seems like a huge performing boost

File: sporkfish/zobrist_hasher.py [hunk 15]
```diff
@@ -1,36 +1,182 @@
...
+    # into an int from [0, 15].
+    castling_keys_idx = 0
+    for idx in range(num_castle_rights):
+        if castling_rights[idx]:
+            castling_keys_idx += 1 << idx
+    return board_hash ^ _castling_keys[castling_keys_idx]  # type: ignore
+
+
+@njit(cache=True, nogil=True)
+def _full_zobrist_hash(
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:79]: numba doesn't interact well with classes. The trick is essentially to create a class level function which then calls the free numba functions

File: tests/test_zobrist_hash.py [hunk 16]
```diff
@@ -1,31 +1,140 @@
...
+        board = BoardFactory.create(BoardPyChess)
+        hash1 = zh1.full_zobrist_hash(board).zobrist_hash
 
-def test_no_hash_collision():
-    zh = ZobristHasher()
+        zh2 = ZobristHasher()
+        hash2 = zh2.full_zobrist_hash(board).zobrist_hash
+        # Make sure internal hashes are equivalent (i.e. we seeded the random numbers) by checking the final hash.
+        # Also check the final hashes are the same for the same board
+        assert hash1 == hash2
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply tests/test_zobrist_hash.py:18]: Should be the same on all platforms

File: sporkfish/zobrist_hasher.py [hunk 17]
```diff
@@ -40,82 +186,93 @@ def __init__(self, board: Board) -> None:
...
+        """
+        Compute the Zobrist hash value incrementally after a move.
 
+        :param board: The chess board.
+        :type board: Board
+        :param move: The move being made.
+        :type move: chess.Move
+        :param prev_state: The previous Zobrist state information.
+        :type prev_state: ZobristStateInfo
+        :param previous_from_square_piece: The piece from the originating square of move.
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:237]: Correct on both

File: sporkfish/zobrist_hasher.py [hunk 18]
```diff
@@ -40,82 +186,93 @@ def __init__(self, board: Board) -> None:
...
-        - __init__(): Initialize the Zobrist Hasher object.
-        - hash(board) -> int: Hashes the given chess board using Zobrist hashing.
-        """
-        # Sporkfish's lucky number
-        np.random.seed(10101010)
+        from_color_piece_type = hash(previous_from_square_piece)
 
-        self._piece_hashes = np.random.randint(0, 2**64, size=(64, 12), dtype=np.uint64)
-        self._turn_hash = np.random.randint(0, 2**64, dtype=np.uint64)
+        # XOR out the previous from square piece
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:247]: XOR is an operation with B XOR A XOR A = B

So if in the previous hash I did B XOR MOVE, and now I add in XOR PREVMOVE they should cancel out if they're the same thing

File: sporkfish/searcher/minimax.py (hunk 6 above)
Comment (KYLChiu) [reply sporkfish/searcher/minimax.py:206]: This potentially is abit dodgy but I see OK performance and moves aren't that bad. Essentially TT retrieves moves with higher depth than previous moves. By treating all quiescence moves at depth 0, if I encounter the position again I essentially skip having it being evaluated regardless of the quiescence depth. This is in contrast to taking the depth to go from quiescence depth 4 to 0 which is essentially 0 to -4 in our negamax depth.

However with how it's evaluated the score is recursively propagated back up so I didn't see the issue in doing it - the "negative depth" score gets evaluated first and then propagated back up to depth 0, so technically if the depth 0 position gets encountered again it's still the same score.

File: sporkfish/transposition_table.py (hunk 8 above)
Comment (KYLChiu) [reply sporkfish/transposition_table.py:24]: The original is correct no?

If not there, overwrite
If there, overwrite if depth > stored depth

Comment (ccjeremylo) [reply sporkfish/transposition_table.py:24]: ahh yes ofc, I was tilting

File: config.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (KYLChiu) [reply config.yml:13]: This is related to the presentation - I mentioned a common pitfall is thinking that in iterative deepening TT is an auto win. But the problem is that the same position in a second iteration of ID is one depth higher now so TT can't be reused

The point of this is to save exactly the same position on the same/similar negamax branch or quiescence or move ordering (to come).

File: sporkfish/zobrist_hasher.py [hunk 19]
```diff
@@ -1,36 +1,182 @@
...
-    Attributes:
-    - colored_piece_types (numpy.ndarray): Array of colored piece types on each square.
-    - board_turn (bool): True if it's black's turn, False otherwise.
-    - en_passant_file (int): File index of the en passant square or -1 if not an en passant square.
-    - castling_rights (numpy.ndarray): Array indicating castling rights.
+    :param zobrist_hash: The Zobrist hash value representing the current board state
+    :type zobrist_hash: np.int64
+    :param ep_file: The file where en passant is possible
+    :type ep_file: np.int8
+    :param castling_rights: An array representing castling rights, defaults to np.array([False, False, False, False])
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:134]: Get back to u, I can't see on phone

File: sporkfish/searcher/minimax.py (hunk 3 above)
Comment (KYLChiu) [reply sporkfish/searcher/minimax.py]: Reverted

File: sporkfish/zobrist_hasher.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py]: Fixed!

File: sporkfish/zobrist_hasher.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py]: I removed the faulty comment!

File: sporkfish/zobrist_hasher.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ccjeremylo) [reply sporkfish/zobrist_hasher.py:78]: I see - I would prefer to put the helper functions in a separate file in that case - just easier on the eyes

Not blocker - can be easily done in a next PR if needs be

File: sporkfish/zobrist_hasher.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ccjeremylo) [reply sporkfish/zobrist_hasher.py:151]: makes sense - keen to understand if this (making function static if possible) is more performant. 

Just curious, not important in terms of this PR.

File: sporkfish/zobrist_hasher.py [hunk 20]
```diff
@@ -40,82 +185,93 @@ def __init__(self, board: Board) -> None:
...
+        # colored_piece_types for all pieces that exist on the board
+        squares_colored_piece_types = np.array(
+            [
+                [square, hash(piece)]
+                for square in chess.SQUARES
+                if (piece := board.piece_at(square))
+            ],
+            dtype=np.int8,
+        )
+        # Splice into two arrays
```
Comment (ccjeremylo) [inline sporkfish/zobrist_hasher.py:207]: Typo? Suspect you mean slice

File: sporkfish/zobrist_hasher.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ccjeremylo) [reply sporkfish/zobrist_hasher.py:246]: Ahh, I suspect this inline comment is referring the line above?

I was confused as I thought the comment is for the two lines below (just putting stuff into a list).

If my understanding is correct - can you move the inline comment up a line plz?

File: sporkfish/zobrist_hasher.py [hunk 21]
```diff
@@ -40,82 +185,93 @@ def __init__(self, board: Board) -> None:
...
 
-        self._piece_hashes = np.random.randint(0, 2**64, size=(64, 12), dtype=np.uint64)
-        self._turn_hash = np.random.randint(0, 2**64, dtype=np.uint64)
+        # XOR out the previous from square piece
+        squares_list = [move.from_square]
+        colored_piece_types_list = [from_color_piece_type]
 
-        self._en_passant_hashes = np.random.randint(0, 2**64, size=8, dtype=np.uint64)
-        self._castling_hashes = np.random.randint(0, 2**64, size=4, dtype=np.uint64)
+        # XOR in the to square piece, piece type depending on if promoted or not
```
Comment (ccjeremylo) [inline sporkfish/zobrist_hasher.py:250]: What's the difference between "XOR in" vs "XOR out"?

Very minor, but would prefer if this line of comment is moved down a line?

**Review by ccjeremylo — COMMENTED** [review]
Couple more very trivial comments

File: sporkfish/zobrist_hasher.py (hunk 21 above)
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:250]: Both are same in terms of mathjematical operation but easier for me to think about. In means its the singularly in the hash whereas out is if theres two copies in the hash (i.e. it cancels out).

File: sporkfish/zobrist_hasher.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:246]: No, putting it into the list allows us to do the numba hash - the hash() function on line 245 is not related to ZH at all (its just to convert the piece type to an number).

File: sporkfish/zobrist_hasher.py (hunk 20 above)
Comment (KYLChiu) [reply sporkfish/zobrist_hasher.py:207]: Oops, yes. Next PR please.

File: sporkfish/searcher/minimax.py (hunk 6 above)
Comment (ccjeremylo) [reply sporkfish/searcher/minimax.py:206]: Hmm interestinggg - my brain is a bit fried atm, need to thinking deeper about this one later

But agree, seems to be working as the tests are not breaking

**Review by ccjeremylo — APPROVED** [review]
LGTM - very well written/designed

Few very minor points on inline comments - can be fixed in next PR - approved 😸

File: config.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (ccjeremylo) [reply config.yml:13]: ahh I rmb now, clearly didn't really understood/digested what you said...

**Pull request merged**